### DISCUSSION TITLE
Provision key to db

### DISCUSSION THREAD

**Discussion by RobertoSnap** [decentralized-identity/veramo#564] [Q&A] [answered]
Hi

Running am Veramo agent on a server. Need to provision the database with a certain privatekey (its currently in uint8array format). 

How do i create a new key (or identifier) from a certain privatekey?

Here is our setup and what we are testing against

https://github.com/BROKLab/auth-bot/blob/veramo/src/network/veramo.service.ts#L71

**Comment by mirceanis** [comment, marked as answer]
Hi, here's an example of importing an Ed25519 signing key:

```ts
const keyData = {
        kid: 'ea75250531f6834328ac210618253288e4c54632962a9708ca82e4a399f79000',
        kms: 'local',
        type: <TKeyType>'Ed25519',
        publicKeyHex: 'ea75250531f6834328ac210618253288e4c54632962a9708ca82e4a399f79000',
        privateKeyHex: '65f341541643070564bb48d9fc10556f2dec246fa056e436a8ec1cdef8c74766ea75250531f6834328ac210618253288e4c54632962a9708ca82e4a399f79000',
      }
const imported = await agent.keyManagerImport(keyData)
```

Is this what you mean by provisioning the database with a key?
If not, then please elaborate

**Reply by RobertoSnap** [reply]
Yes, exactly. Thanks

**Reply by RobertoSnap** [reply]
Or wait, that would not create the identifier i assume.

Would it be possible to expose an interface that just takes pk hex? Preferable on the didManager? 😬

**Reply by mirceanis** [reply]
That is true, it would not create an identifier, only a keypair.
But, the same import logic can be used to import an identifier. It's just a bit inconvenient at the moment.
Of course, a helper method to bootstrap an identifier directly from an imported key would be a nice addition and a moderately easy [`good-first-issue`](https://github.com/uport-project/veramo/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22)

**Reply by RobertoSnap** [reply]
Trying to do 

`
 console.log('TEST', keyData.privateKeyHex);

    return await this.agent.didManagerImport({
      services: [],
      provider: 'did:key',
      did: keyData.did,
      controllerKeyId: keyData.kid,
      keys: [
        {
          kid: keyData.kid,
          kms: 'local',
          type: <TKeyType>'Ed25519',
          publicKeyHex: keyData.publicKeyHex,
          privateKeyHex: keyData.privateKeyHex,
        },
      ],
    });
`

Which gives me correct identifier table


did | provider | alias | saveDate | updateDate | controllerKeyId
-- | -- | -- | -- | -- | --
did:key:z6MkwLAeA71m4h7qen6a1yQrk17usWJJyvTFGjG24EpG2fW5 | did:key | NULL | 2021-06-17 10:16:04 | 2021-06-17 10:16:04 | fac58071979cb1b51f6766f1de9475103e439831b727809499887af90d194826



But keytable has a weird long privatekey


kid | kms | type | publicKeyHex | privateKeyHex | meta | identifierDid
-- | -- | -- | -- | -- | -- | --
fac58071979cb1b51f6766f1de9475103e439831b727809499887af90d194826 | local | Ed25519 | fac58071979cb1b51f6766f1de9475103e439831b727809499887af90d194826 | f069657d1533920484a429cbf275961d230a945814f15ac709083269826434467be9f643bbd392a5a9fb61c5e598cdcd4d200e4d3591a06f3699d8fc3dae56f12c0fd0bfc0fb3ac09df591d88219c03920c0cc6bd8ddd1b12e68e7eac019bbc98b2abe20d520fa1229cd08bee17d4566e2453e5f84c05229f6f2e9801a37e2c1ce22858ff33af2f180b656627981aec6ed6847bad5eb6382b92cfb16ecec4c2ed005d06f4ff14785 | NULL | did:key:z6MkwLAeA71m4h7qen6a1yQrk17usWJJyvTFGjG24EpG2fW5

THe private key that is issued to import is 050c6546d13df26532bde557562efe88a8544205bd1dd804e1aafcdbbd20616dfac58071979cb1b51f6766f1de9475103e439831b727809499887af90d194826

Other variables issues to import
`ISSUER_DID="did:key:z6MkwLAeA71m4h7qen6a1yQrk17usWJJyvTFGjG24EpG2fW5"
ISSUER_KID="fac58071979cb1b51f6766f1de9475103e439831b727809499887af90d194826"
ISSUER_PUBLIC_KEY_HEX="fac58071979cb1b51f6766f1de9475103e439831b727809499887af90d194826"
ISSUER_PRIVATE_KEY_HEX="050c6546d13df26532bde557562efe88a8544205bd1dd804e1aafcdbbd20616dfac58071979cb1b51f6766f1de9475103e439831b727809499887af90d194826"`

**Reply by mirceanis** [reply]
that is expected. the private key in the database table is encrypted using the `SecretBox` you provide during setup.

**Reply by RobertoSnap** [reply]
aaa 👍

**Reply by mirceanis** [reply]
I'm sure this need not be said but, just in case it's not already obvious, don't reuse these keys except for tests.

**Reply by RobertoSnap** [reply]
hehe, its obvious, this is testing env :)

**Reply by RobertoSnap** [reply]
When useing didManagerImport to import .keys

Should they be provided encrypted with secret box or un-ecrypted? 

I assumed it was encrypted when useing secretbox and un-encrypted when useing secret box( ofc with same secret box entropy).

**Reply by mirceanis** [reply]
When providing the data to didManagerImport, it should be in clear text form.
